rfcs/table-destructuring.md

```diff
@@ -0,0 +1,236 @@
+    blue = "best";
+}
+
+local red, green, blue = foo.red, foo.green, foo.blue
+print(red) -- good
+print(green) -- better
+print(blue) -- best
+```
+```lua
+-- New
+local foo = {
+    red = "good";
+    green = "better";
+    blue = "best";
+}
+
+local {red, green, blue} = foo
+print(red) -- good
+print(green) -- better
+print(blue) -- best
+```
+
+### Variable Assignments for Function Returns
+```lua
+-- Current
+local function f()
+    return {
+        name = "John";
+        age = 25;
+        gender = "male";
+    }
+end
+
+local data = f()
+local name, age, gender = data.name, data.age, data.gender
+
+print(name) -- John
+print(age) -- 25
+print(gender) -- male
+```
+```lua
+-- New
+local function f()
+    return {
+        name = "John";
+        age = 25;
+        gender = "male";
+    }
+end
+
+local {name, age, gender} = f()
+print(name) -- John
+print(age) -- 25
+print(gender) -- male
+```
+
+### Variable Assignments for UserData & Vectors
+```lua
+-- Current
+local Part = workspace.Part
+local Position = Part.CFrame.Position
+local X, Y, Z = Position.X, Position.Y, Position.Z
+
+print(X, Y, Z) -- 0, 0, 0
+```
+```lua
+-- New
+local Part = workspace.Part
+local {Position} = Part.CFrame
+local {X, Y, Z} = Position
+
+print(X, Y, Z) -- 0, 0, 0
+```
+
+### Variable Assignments for Indexing Children (???)
+```lua
+-- Current
+local Baseplate = workspace.Baseplate
+print(Baseplate) -- Baseplate
+```
+```lua
+-- New
+local {Baseplate} = workspace
+print(Baseplate) -- Baseplate
+```
+
+---
+
+### Destructuring Function Parameters
+```lua
+-- Before
+local function f(data)
+    local name = data.name
+    local age = data.age
+    local gender = data.gender
+
+    print(name, age, gender) -- John, 25, male
+end
+
+f({
+    name = "John";
+    age = 25;
+    gender = "male";
+})
+```
+```lua
+-- New
+local function f({ name, age, gender })
+    print(name, age, gender) -- John, 25, male
+end
+
+f({
+    name = "John";
+    age = 25;
+    gender = "male";
+})
+```
+
+---
+
+### Defining Alternative Identifiers (Questionable?)
+In this example, `fizz-buzz` is not a valid identifier. You should be able to instead define an alternative identifier, like so.
+```lua
+local foo = {
+    ["fizz-buzz"] = true;
+}
+
+local { fizzbuzz = "fizz-buzz" } = foo
+print(fizzbuzz) -- fizz-buzz
+```
+
+### More Complex Example
+This example is a direct port of the JavaScript example I gave on issue #617.
+```lua
+local obj = {
+    name = "John";
+    records = {
+        {
+            date = "7/29/2022";
+            text = "Hello World!";
+        }
+    }
+}
+
+local {
+    name,
+    records = {
+        { text }
+    }
+} = obj
+
+print(name .. " says " .. text) -- John says Hello World!
+```
+
+## Drawbacks
+
+From my perspective, this would be a rather large addition to the language. *Some* of the drawbacks are listed below:
+* Increased compiler complexity
+* Considered language bloat
+* New, unknown syntax
+* Not necessarily friendly to new programmers
```

Comment:
If the RFC tried to follow an existing syntax with some changes for type syntax (which is unmentioned in this RFC), it would be helpful and would make the language more familiar to programmers coming from languages which have similar ways of destructuring. If a new syntax is going to be used it would be helpful to explain why certain choices are being made.